### PR TITLE
fix(swarm): add dnsaddr support in swarm connect

### PR DESCRIPTION
**Pull request by overbool**
Fixes: https://github.com/ipfs/go-ipfs/issues/5522

License: MIT
Signed-off-by: Overbool <overbool.xu@gmail.com>

### CODE REVIEW

File: core/commands/swarm.go [hunk 1]
```diff
@@ -557,10 +560,43 @@ func parseAddresses(addrs []string) (iaddrs []iaddr.IPFSAddr, err error) {
...
-	iaddrs, err := parseAddresses(addrs)
+	// resolve the addrs
+
+	var maddrs []ma.Multiaddr
+	for _, addr := range addrs {
+		maddr, err := ma.NewMultiaddr(addr)
+		if err != nil {
+			return nil, err
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (michaelavila) [inline core/commands/swarm.go]: I've looked through this whole PR and everything makes sense to me :+1:. I do have a question about how these timeouts are specified in go-ipfs (both here and more generally). I notice lines like the following (and there are a handful):

https://github.com/ipfs/go-ipfs/blob/master/core/bootstrap.go#L60

But also lines like yours (e.g.):

https://github.com/ipfs/go-ipfs/blob/master/core/commands/pubsub.go#L178

Is there a rule of thumb for deciding when to specify these timeouts with a name vs inline?

Comment (Stebalien) [reply core/commands/swarm.go]: Well, we should probably name this. We just get lazy sometimes.

File: core/commands/swarm.go [hunk 2]
```diff
@@ -557,10 +560,43 @@ func parseAddresses(addrs []string) (iaddrs []iaddr.IPFSAddr, err error) {
...
+		raddrs, err := madns.Resolve(ctx, maddr)
```
Comment (Stebalien) [inline core/commands/swarm.go]: Can we parallelize this?

File: core/commands/swarm.go [hunk 3]
```diff
@@ -557,10 +560,43 @@ func parseAddresses(addrs []string) (iaddrs []iaddr.IPFSAddr, err error) {
...
+		cancel()
+		if err != nil {
+			return nil, err
+		}
+		if len(raddrs) == 0 {
+			return nil, fmt.Errorf("non-resolvable multiaddr about %v", maddr)
+		}
+		maddrs = append(maddrs, raddrs...)
```
Comment (Stebalien) [inline core/commands/swarm.go]: I believe we may need to filter out addresses that *still* don't end in `/ipfs/Qm...`.

**Comment by hsanjuan** [comment]
> we'll insert the resolved addrs into the peerstore instead of the dnsaddr.

Maybe this has changed, but do unresolved addresses in the peerstore get resolved when building a connection? My impression when I was testing this a while ago is they didn't. I remember a discussion or PR in which @lanzafame was involved (??) suggesting to do this at a lower level in libp2p.

A few months back, addresses were only resolved in `basichost.Connect` (https://github.com/libp2p/go-libp2p/blob/master/p2p/host/basic/basic_host.go#L383) but normally things will use `NewStream` directly and I think (i might be wrong) that that codepath does not resolve if it has to open a new connection.

So might be better off storing resolved addresses for the moment?

**Comment by Stebalien** [comment]
Ah... No, they don't. That's broken. Yeah, you're right. But we should *really* fix this. We don't even do routing on that path.

**Comment by overbool** [comment]
>For each address that doesn't end in /ipfs/Qm..., resolve it.

@Stebalien Did you mean that if address ends in `/ipfs/Qm...`, we don't need to resolve it, because it will resolve address inner the `basichost.Connect`?

**Comment by overbool** [comment]
@Stebalien Thx your advice, i had fixed it.

**Pull request merged**